File: p013e/roman_to_int.js
/**
 * @param {string} s
 * @return {number}
 */
const VALUE_MAP = {
  I: 1,
  V: 5,
  X: 10,
  L: 50,
  C: 100,
  D: 500,
  M: 1000,
};

var romanToInt = function (s) {
  let ans = 0;
  for (i = 0; i < s.length; ++i) {
    let [curr, next] = [VALUE_MAP[s[i]], VALUE_MAP[s[i + 1]] || 0];
    if (curr < next) ans -= curr;
    else ans += curr;
  }
  return ans;
};
// TESTS
[
  ['III', 3],
  ['IV', 4],
  ['VII', 7],
  ['IX', 9],
  ['LVIII', 58],
  ['XCIX', 99],
  ['DCCCXC', 890],
  ['MCMXCIV', 1994],
].forEach(([s, expected]) => {
  const actual = romanToInt(s);
  console.log('Roman', s, 'to int ->', actual);
  console.assert(actual === expected);
});

File: p013e/roman_to_int.py
class Solution:
    def romanToInt(self, s: str) -> int:
        VALUE_MAP = {
            "I": 1,
            "V": 5,
            "X": 10,
            "L": 50,
            "C": 100,
            "D": 500,
            "M": 1000,
        }
        ans = 0
        for i in range(0, len(s) - 1):
            cur = VALUE_MAP[s[i]]
            if cur < VALUE_MAP[s[i + 1]]:
                ans -= cur
            else:
                ans += cur
        return ans + VALUE_MAP[s[-1]]


# TESTS
for s, expected in [
    ["III", 3],
    ["IV", 4],
    ["VII", 7],
    ["IX", 9],
    ["LVIII", 58],
    ["XCIX", 99],
    ["DCCCXC", 890],
    ["MCMXCIV", 1994],
]:
    sol = Solution()
    actual = sol.romanToInt(s)
    print("Roman", s, "to int ->", actual)
    assert actual == expected
